sycl/include/sycl/ext/oneapi/reduction.hpp

```diff
@@ -2500,30 +2447,24 @@ reduSaveFinalResultToUserMemHelper(std::vector<event> &,
                                    std::shared_ptr<detail::queue_impl>, bool) {}
 
 template <typename Reduction, typename... RestT>
-std::enable_if_t<Reduction::is_usm>
-reduSaveFinalResultToUserMemHelper(std::vector<event> &Events,
-                                   std::shared_ptr<detail::queue_impl> Queue,
-                                   bool IsHost, Reduction &, RestT... Rest) {
-  // Reductions initialized with USM pointer currently do not require copying
-  // because the last kernel write directly to USM memory.
-  reduSaveFinalResultToUserMemHelper(Events, Queue, IsHost, Rest...);
-}
-
-template <typename Reduction, typename... RestT>
-std::enable_if_t<!Reduction::is_usm> reduSaveFinalResultToUserMemHelper(
+void reduSaveFinalResultToUserMemHelper(
     std::vector<event> &Events, std::shared_ptr<detail::queue_impl> Queue,
     bool IsHost, Reduction &Redu, RestT... Rest) {
-  if (Redu.hasUserDiscardWriteAccessor()) {
-    handler CopyHandler(Queue, IsHost);
-    auto InAcc = Redu.getReadAccToPreviousPartialReds(CopyHandler);
-    auto OutAcc = Redu.getUserDiscardWriteAccessor();
-    Redu.associateWithHandler(CopyHandler);
-    if (!Events.empty())
-      CopyHandler.depends_on(Events.back());
-    CopyHandler.copy(InAcc, OutAcc);
-    event CopyEvent = CopyHandler.finalize();
-    Events.push_back(CopyEvent);
-  }
+  // Reductions initialized with USM pointer currently do not require copying
+  // because the last kernel writes directly to the USM memory.
+  if constexpr (!Reduction::is_usm)
```

Comment:
Personally I would prefer to have curly-brackets around the body of this because it is multi-line. Not a big requirement though.